backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java

```diff
@@ -382,18 +390,122 @@ public CommitEntity getCommitEntity(
+        } else {
+          if (repositoryIdentification.getRepoId() != newRepoId) {
+            throw new ModelDBException(
+                "DatasetVersion '"
+                    + commitEntity.getCommit_hash()
+                    + "' associated with multiple datasets",
+                Code.INTERNAL);
+          }
+        }
+
+        if (repositoryEntity == null) {
+          repositoryEntity =
+              repositoryDAO.getRepositoryById(session, repositoryIdentification, true, false);
+        }
+
+        Query query = session.createQuery(RepositoryDAORdbImpl.CHECK_BRANCH_IN_REPOSITORY_HQL);
+        query.setParameter("repositoryId", repositoryEntity.getId());
+        query.setParameter("branch", ModelDBConstants.MASTER_BRANCH);
+        BranchEntity branchEntity = (BranchEntity) query.uniqueResult();
+
+        session.beginTransaction();
+        CommitEntity parentDatasetVersion =
+            new ArrayList<>(commitEntity.getParent_commits()).get(0);
+
+        if (branchEntity != null
+            && branchEntity.getCommit_hash().equals(commitEntity.getCommit_hash())) {
+          repositoryDAO.setBranch(
+              SetBranchRequest.newBuilder()
+                  .setRepositoryId(repositoryIdentification)
+                  .setBranch(ModelDBConstants.MASTER_BRANCH)
+                  .setCommitSha(parentDatasetVersion.getCommit_hash())
+                  .build(),
+              false);
+        }
+
+        if (!commitEntity.getChild_commits().isEmpty()) {
+          List<CommitEntity> childCommits = new ArrayList<>(commitEntity.getChild_commits());
+          String updateChildEntity =
+              "UPDATE commit_parent cp SET cp.parent_hash = :parentHash WHERE cp.child_hash = :childHash";
+          Query updateChildQuery = session.createSQLQuery(updateChildEntity);
+          updateChildQuery.setParameter("parentHash", parentDatasetVersion.getCommit_hash());
+          updateChildQuery.setParameter("childHash", childCommits.get(0).getCommit_hash());
+          updateChildQuery.executeUpdate();
+        }
 
-      String entityName = "Commit";
-      if (isDatasetVersion) {
-        entityName = "DatasetVersion";
+        if (commitEntity.getRepository().size() == 1) {
+          String compositeId =
+              VersioningUtils.getVersioningCompositeId(
+                  repositoryEntity.getId(),
+                  commitEntity.getCommit_hash(),
+                  Collections.singletonList(ModelDBConstants.DEFAULT_VERSIONING_BLOB_LOCATION));
+          DeleteEntitiesCron.deleteLabels(
+              session, compositeId, IDTypeEnum.IDType.VERSIONING_REPO_COMMIT_BLOB);
+          DeleteEntitiesCron.deleteAttribute(session, compositeId);
+          DeleteEntitiesCron.deleteTagEntities(
+              session, repositoryEntity.getId(), commitEntity.getCommit_hash());
+          session.delete(commitEntity);
+        } else {
+          commitEntity.getRepository().remove(repositoryEntity);
```

Comment:
when will this happen for dataset? throw an exception?

Reply:
We are throwing exception from above code `if (commitEntity.getRepository() != null && commitEntity.getRepository().size() > 1) {` so this else never call. i have removed it.